apps/prairielearn/src/models/course-instances.ts

```diff
@@ -22,22 +24,58 @@ const CourseInstanceAuthzSchema = CourseInstanceSchema.extend({
 });
 export type CourseInstanceAuthz = z.infer<typeof CourseInstanceAuthzSchema>;
 
-export async function selectCourseInstanceById(
-  course_instance_id: string,
-): Promise<CourseInstance> {
-  return queryRow(sql.select_course_instance_by_id, { course_instance_id }, CourseInstanceSchema);
+export async function selectCourseInstanceById({
+  id,
+  requestedRole,
+  authzData,
+}: {
+  id: string;
+  requestedRole: 'System' | 'Student' | 'Student Data Viewer' | 'Student Data Editor' | 'Any';
+  authzData: AuthzData;
+}) {
+  const courseInstance = await queryRow(
+    sql.select_course_instance_by_id,
+    { course_instance_id: id },
+    CourseInstanceSchema,
+  );
+
+  if (courseInstance.share_source_publicly) {
+    return courseInstance;
+  }
```

Comment:
Discussed with @reteps: this is broken in its current implementation. The intended meaning of this function is "give it a role, return course instance if user has that role, otherwise throw a 403 error". This short-circuit breaks that.